### ISSUE TITLE
BUG? Undefined? behavior with  --cache-tmp-upload-path  cache/vfs

### ISSUE THREAD

**Issue by abclution** [rclone/rclone#4226]
#### What is the problem you are having with rclone?
Undefined? behavior with  --cache-tmp-upload-path

The back-end is a crypt on top of cache on top of Google Drive. Which I believe is the recommended way to do it?

When using --vfs-cache-mode write or full, without setting --cache-dir  or -cache-tmp-upload-path,  write caching doesn't work, but everything else works like normal. This is expected and this is OK.

`/usr/bin/rclone mount GD-TD-ABC-Cached-Xcrypt: /mypool/mount --daemon --fast-list --vfs-cache-mode full --dir-cache-time 60m --allow-other --allow-non-empty -`

When adding  -cache-tmp-upload-path=  (no --cache-dir), write  caching begins to work but not as expected.

If I use this command: 

`/usr/bin/rclone mount GD-TD-ABC-Cached-Xcrypt: /mypool/mount --cache-tmp-upload-path=~/fartsOnWheels --daemon --fast-list --vfs-cache-mode full --dir-cache-time 60m --allow-other --allow-non-empty -`

- **If  ~/fartsOnWheels does not exist.**
The write caching begins to work.
The log mentions the given directory.  
LOG:  Upload Temp FS: ~/fartsOnWheels
But the cache-tmp-upload directory of ~/fartsOnWheels is not obeyed and not created.
- **If  ~/fartsOnWheels does exist.**
The write caching begins to work.
The log mentions the given directory.  
 LOG:  Upload Temp FS: ~/fartsOnWheels
But the cache-tmp-upload directory of ~/fartsOnWheels is not obeyed
- **If  instead I give a full path /home/user/fartsOnWheels , it is created if it doesnt exist, otherwise it uses the existing and ...**
The write caching begins to work.
The log mentions the given directory.  
 LOG:   Upload Temp FS: /home/user/fartsOnWheels 
And the cache-tmp-upload directory of /home/user/fartsOnWheels is obeyed and fully functional.

**_In both first two cases, write caching begins to work, but the temp upload cache dir is automatically instead /home/user/.cache/rclone/vfs/GD-TD-ABC-Cached-Xcrypt/ (part of the vfs cache system!)_**

Now it turns out that the issue is because of not using full paths and instead using the (tilde glob) ~ character to indicate the user home folder.
In any case by giving a full path /home/user/fartsOnWheels/ the directly is correctly used.

**But the plot thickens**, if I put literally anything that isn't a full path, for example --cache-tmp-upload-path=yes it seems to fall back to the same behavior where the temp upload cache dir is /home/user/.cache/rclone/vfs/GD-TD-ABC-Cached-Xcrypt/

**Something else to note, while this undocumented automatic fallback enables the write cache to function, the files copied during this time for write, while in the undocumented fallback temp upload folder, are never encrypted (or names scrambled) while in this folder, although when they are uploaded fully on the destination they are properly encrypted.**

Example: File copied to the mount for upload, with the fallback temp active.
--cache-tmp-upload-path=~/fartsOnWheels 

```
 ls -la ~/.cache/rclone/vfs/GD-TD-ABC-Cached-Xcrypt/
 2013-cvma-scientific-presentation.pdf
```
(The file name is not scrambled in the temp staging area on disk)

LOG: INFO  : 6hr4ip2k9c481h9lhmu62tvqm2i4t8pikupd9iqpajfevm4kg49gb3nsb3v5qm9iq1r3e1jce4l3a: background upload: started upload

 In contrast to a more correctly working full path assigned for  --cache-tmp-upload-path=~/fartsOnWheels

Example: File copied to the mount for upload, with full path --cache-tmp-upload-path=~/fartsOnWheels

```
ls /tmp/fartsonwheels/
6hr4ip2k9c481h9lhmu62tvqm2i4t8pikupd9iqpajfevm4kg49gb3nsb3v5qm9iq1r3e1jce4l3a
```
(The file name is scrambled in the temp staging area on disk)

INFO  : 6hr4ip2k9c481h9lhmu62tvqm2i4t8pikupd9iqpajfevm4kg49gb3nsb3v5qm9iq1r3e1jce4l3a: background upload: started upload

In either case, the file is uploaded, name scrambled and encrypted correctly on the destination after.

So, quite honestly, the fact that the tilde doesn't seem to work for the path for the --cache-tmp-upload-path= could be due to many things, perhaps I need to put it in quotations? I don't know.

To be quite honest, the undefined behavior is actually closer to what I was trying to accomplish in the end..
I would prefer that a sane default for cache-tmp-upload-path would be the default, either in the cache-backend folder, or the vfs folder. And by using the tilde glob I was actually trying to put the staging area somewhere in the nearby area a.k.a the next best thing.

So, what do I wish would happen?
1. Would be great if the tilde glob was obeyed.
2. A sane default and documented temp upload area based on the defaults for the other cache methods would be great too.

Hope this made sense, it was quite difficult to figure out what was going on and difficult to explain it.

#### What is your rclone version (output from `rclone version`)

rclone v1.51.0
- os/arch: linux/amd64
- go version: go1.13.7

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Debian 10 Buster

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
`/usr/bin/rclone mount GD-TD-ABCl-Cached-Xcrypt: /mypool/mount --cache-tmp-upload-path=literallyanythinghere --daemon --fast-list --vfs-cache-mode full --dir-cache-time 60m --allow-other --allow-non-empty -`

**Comment by animosity22** [comment]
If you can, please use the forums first for a question before posting an issue as stated in the issue template.

The VFS backend and the cache backend are two different backends.

cache-tmp-upload goes with the cache backend and is used for the cache backend as a place to write files.

vfs-cache-mode writes is part of the standard VFS backend and is used to have some write capability that's needed for certain write operations.

With the debug log on, you can see which backend is being used for writing based on your config.

The tilde does work, but nothing generally recommended:

```
felix@gemini:/opt/rclone$ rclone mount cachecrypt: /home/felix/test --cache-tmp-upload-path ~/logs -vv
2020/05/13 06:33:37 DEBUG : rclone: Version "v1.51.0" starting with parameters ["rclone" "mount" "cachecrypt:" "/home/felix/test" "--cache-tmp-upload-path" "/home/felix/logs" "-vv"]
```

and you can see in the logs

```
2020/05/13 06:33:38 INFO  : cache: Upload Temp FS: /home/felix/logs
```

as an example as that answers your #1.

I'm not sure what you are asking for #2.

**Comment by abclution** [comment]
> If you can, please use the forums first for a question before posting an issue as stated in the issue template.

This is a bug report. Undefined behavior is either a bug or a feature.

> 
> The VFS backend and the cache backend are two different backends.

Yes I know this.

> 
> cache-tmp-upload goes with the cache backend and is used for the cache backend as a place to write files.
> 

Yes I know this as well.

> vfs-cache-mode writes is part of the standard VFS backend and is used to have some write capability that's needed for certain write operations.
> 

Yup.
> With the debug log on, you can see which backend is being used for writing based on your config.
> 
As I tried to explain, perhaps poorly, the log does mention the cache-tmp-upload-path folder, but it DOES NOT use it. It uses the vfs folder.

> The tilde does work, but nothing generally recommended:
> 
> ```
> felix@gemini:/opt/rclone$ rclone mount cachecrypt: /home/felix/test --cache-tmp-upload-path ~/logs -vv
> 2020/05/13 06:33:37 DEBUG : rclone: Version "v1.51.0" starting with parameters ["rclone" "mount" "cachecrypt:" "/home/felix/test" "--cache-tmp-upload-path" "/home/felix/logs" "-vv"]
> ```
> 
> and you can see in the logs
> 
> ```
> 2020/05/13 06:33:38 INFO  : cache: Upload Temp FS: /home/felix/logs
> ```
> 
> as an example as that answers your #1.
> 
> I'm not sure what you are asking for #2.

`2020/05/13 13:44:58 DEBUG : rclone: Version "v1.51.0" starting with parameters ["/usr/bin/rclone" "mount" "GD-TD-ABClution-C-X:" "/mypool/CloudStorages/GD_abclution_TD" "--cache-tmp-upload-path" "~/fartsOnWheels" "--daemon" "--fast-list" "--vfs-cache-mode"            "full" "--dir-cache-time" "60m" "--allow-other" "--allow-non-empty" "--drive-acknowledge-abuse=true" "--user-agent" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36" "--log-level" "DEBUG" "--log-fil           e=/mypool/CloudStorages/GD-TD-ABClution-C-X.log"]
2020/05/13 13:44:58 DEBUG : Using config file from "/home/user/.config/rclone/rclone.conf"
2020/05/13 13:44:59 DEBUG : GD-TD-ABClution-C: wrapped GD-TD-ABClution:SAFESTORAGE at root SAFESTORAGE
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Cache DB path: /home/user/.cache/rclone/cache-backend/GD-TD-ABClution-C.db
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Cache chunk path: /home/user/.cache/rclone/cache-backend/GD-TD-ABClution-C
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Memory: true
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Size: 512k
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Total Size: 128G
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Clean Interval: 1m0s
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Workers: 4
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: File Age: 6h0m0s
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Cache Writes: enabled
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Upload Temp Rest Time: 15s
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Upload Temp FS: ~/fartsOnWheels
2020/05/13 13:44:59 DEBUG : Adding path "cache/expire" to remote control registry
2020/05/13 13:44:59 DEBUG : Adding path "cache/stats" to remote control registry
2020/05/13 13:44:59 DEBUG : Adding path "cache/fetch" to remote control registry
2020/05/13 13:44:59 DEBUG : Cache remote GD-TD-ABClution-C:SAFESTORAGE: Services stopped
2020/05/13 13:44:59 DEBUG : GD-TD-ABClution-C: wrapped GD-TD-ABClution:SAFESTORAGE at root SAFESTORAGE
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Cache DB path: /home/user/.cache/rclone/cache-backend/GD-TD-ABClution-C.db
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Cache chunk path: /home/user/.cache/rclone/cache-backend/GD-TD-ABClution-C
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Memory: true
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Size: 512k
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Total Size: 128G
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Chunk Clean Interval: 1m0s
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Workers: 4
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: File Age: 6h0m0s
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Cache Writes: enabled
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Upload Temp Rest Time: 15s
2020/05/13 13:44:59 INFO  : GD-TD-ABClution-C: Upload Temp FS: ~/fartsOnWheels
2020/05/13 13:45:00 DEBUG : Adding path "cache/expire" to remote control registry
2020/05/13 13:45:00 DEBUG : Adding path "cache/stats" to remote control registry
2020/05/13 13:45:00 DEBUG : Adding path "cache/fetch" to remote control registry
2020/05/13 13:45:00 DEBUG : Encrypted drive 'GD-TD-ABClution-C-X:': Mounting on "/mypool/CloudStorages/GD_abclution_TD"
`
After copying a file for upload, doing an ls in the cache-tmp-upload-path shows nothing, looking into the vfs default cache location shows the file.


```
user@SPROUT /m/C/scripts> ls ~/fartsOnWheels/
user@SPROUT /m/C/scripts> ls ~/.cache/rclone/vfs/GD-TD-ABClution-C-X/pm1.abclution.com/cdrviewersetup.exe
/home/user/.cache/rclone/vfs/GD-TD-ABClution-C-X/pm1.abclution.com/cdrviewersetup.exe

```

If I run the same rclone command with full path, everything works as expected.

**Comment by animosity22** [comment]
Here is an example mount:

```
felix     385101  375621 19 06:54 pts/0    00:00:23 rclone mount cachecrypt: /home/felix/test --cache-tmp-upload-path /home/felix/logs -vv
```

You can see the tilde gets expanded to the full path and also in the debug log so it really does not matter.

A write goes to the cache-tmp-upload path as seen in an example:

```
2020/05/13 06:56:05 DEBUG : &{jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv (w)}: >Flush: err=<nil>
2020/05/13 06:56:05 DEBUG : &{jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv (w)}: Release:
2020/05/13 06:56:05 DEBUG : jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv: WriteFileHandle.Release nothing to do
2020/05/13 06:56:05 DEBUG : &{jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv (w)}: >Release: err=<nil>
2020/05/13 06:56:15 INFO  : acc9lv0c13ccdr2i14tm8qu4nisnphfijbtisdo9a9qsfhqk7q87udn8eq28np7gutkk138mchitg: background upload: started upload
2020/05/13 06:56:15 DEBUG : acc9lv0c13ccdr2i14tm8qu4nisnphfijbtisdo9a9qsfhqk7q87udn8eq28np7gutkk138mchitg: Need to transfer - File not found at Destination
2020/05/13 06:56:17 DEBUG : acc9lv0c13ccdr2i14tm8qu4nisnphfijbtisdo9a9qsfhqk7q87udn8eq28np7gutkk138mchitg: Sending chunk 0 length 1073741824
2020/05/13 06:56:32 DEBUG : acc9lv0c13ccdr2i14tm8qu4nisnphfijbtisdo9a9qsfhqk7q87udn8eq28np7gutkk138mchitg: Sending chunk 1073741824 length 431578782
```

and the file itself on disk.

```
felix@gemini:~/logs$ ls -ltr | tail -1
-rw-rw-r-- 1 felix felix 1505320606 May 13 06:55 acc9lv0c13ccdr2i14tm8qu4nisnphfijbtisdo9a9qsfhqk7q87udn8eq28np7gutkk138mchitg
```

You need to make a debug log of the copy happening as it's using the vfs writes config rather than the cache tmp upload as you are mixing the two in a config as you are using vfs-cache-mode full along with the cache backend.

**Comment by abclution** [comment]
Thanks for your answer.
Yes, I noticed that the tilde is NOT being expanded properly in my log, this is part of the problem/solution.

But the undefined behavior is what worries me.

The undefined behavior is that rclone is choosing a folder, automatically, within the ~/.cache/rclone/vfs/ folder, for the vfs layer --cache-dir option, which is not set/defined and only the --cache-tmp-upload-path is defined and invalid (due to non expanded glob etc.)

Additionally this automatically chosen folder, files passing through there the file names are left unencrypted, vs when using a valid -cache-tmp-upload-path location, the file names are properly scrambled on disk.

In both cases files are uploaded correctly.

**Comment by animosity22** [comment]
That's the default location for cache-dir if you don't put a path in. Since you have vfs-cache-mode  with something on than off, it uses cache-dir to store anything for that mode as described here:

https://rclone.org/commands/rclone_mount/#file-caching

If you do not define it, you'll see a line with the default value:

```
2020/05/13 07:19:57 DEBUG : Cache remote cache:crypt: subscribing to ChangeNotify
2020/05/13 07:19:57 DEBUG : vfs cache root is "/home/felix/.cache/rclone/vfs/cachecrypt"
2020/05/13 07:19:57 DEBUG : Adding path "vfs/forget" to remote control registry
```

as an example.

If your item is not expanding, that's normally due to permission issues or something along those lines.

Here is a line from my debug log with a bad path.

```
2020/05/13 07:23:06 INFO  : cache: Upload Temp FS: ~joe/BLAH
```

In general, it's not good practice to use blobs there and just use a full path.

**Comment by abclution** [comment]
Ok. That was helpful. This resolves some of the discrepancies.
I did check the doc, but it actually doesn't say where that the default vfs cache folder is.. only how to determine it, which was helpful.

`If run with -vv rclone will print the location of the file cache. The files are stored in the user cache file area which is OS dependent but can be controlled with --cache-dir or setting the appropriate environment variable.`

Also I did see YOU (I assume) answering on the forum the same question here: https://forum.rclone.org/t/where-is-rclone-vfs-cache-dir-by-default/11309

So, that explains part of it.

I was confused about 
--cache-dir (has defaults, functions with defaults if VFS is enabled) 
and
 --cache-tmp-upload-path (no defaults, if not set doesn't function.) 

The rest of the behavior is explained by the existing defaults and bad path. 

Thank you for your help and clarification, appreciate it..
